### PR TITLE
Cleanup azure

### PR DESCRIPTION
**Pull request by nvaytet**
This PR makes heavy use of parameters and templates in azure `.yml` files to reduce duplication between the different pipelines (PR, main, release).

The idea is to use a bash script to perform the build, rather than write the steps in the yml files.
The script is run directly by the builder, and `conda build` is then used to copy the install target files over and create the package.
We do not use conda build to perform the build directly because this makes it impossible to use `ccache`.

We also only upload packages to anaconda once everything has passed, including the documentation build.
Documentation is deployed to github pages only after the conda packages were successfully uploaded to anaconda.

Dry run with building py3.7 and 3.8 for linux, osx and win, as well as docs, and uploading conda packages is here: https://dev.azure.com/scipp/scipp/_build/results?buildId=4363&view=results
The only thing I haven't checked yet is publishing the built docs to github pages, but in principle, all that code was just copy/pasted.

### CODE REVIEW

File: .azure-pipelines/pull_request.yml
Comment (jl-wynen) [inline .azure-pipelines/pull_request.yml:22]: Does it make sense to move the Mac and Windows builds into template files, too?

File: .azure-pipelines/pull_request.yml
Comment (jl-wynen) [inline .azure-pipelines/pull_request.yml:167]: Missing newline. Does this matter?

File: .azure-pipelines/pull_request.yml
Comment (nvaytet) [reply .azure-pipelines/pull_request.yml:19]: Yes, that's the goal. Please hold off with revieweing for now. PR is opened only so I can try the azure builds.

File: .azure-pipelines/templates/build.yml
Comment (OwenArnold) [inline .azure-pipelines/templates/build.yml:60]: Minor point: Env name will not match actual file that produced it, maybe that's confusing if someone then wonders why `scipp-developer-minimal` is using python 3.8 say, but that is not specified for `scipp-developer-minimal.yml`?

File: .azure-pipelines/templates/build.yml
Comment (OwenArnold) [inline .azure-pipelines/templates/build.yml:62]: Could we write the python version into the displayName?

File: .azure-pipelines/templates/build.yml
Comment (OwenArnold) [inline .azure-pipelines/templates/build.yml:68]: Isn't this also "installing"?

File: .azure-pipelines/templates/build.yml
Comment (OwenArnold) [inline .azure-pipelines/templates/build.yml:77]: 👍 Probably is a good idea to start using `--no-anaconda-upload` explicitly

File: .azure-pipelines/templates/ccache.yml
Comment (OwenArnold) [inline .azure-pipelines/templates/ccache.yml:3]: We are going to get this to work for OSX too? I think we could get this from `conda`

**Review by OwenArnold — COMMENTED** [review]
Thanks for the work thus far. This is looking better and easier to extend that what we had before.

File: .azure-pipelines/templates/build.yml
Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:16]: I think in the case of `OSX_VERSION` it has to be left blank, because we test further down the pipeline whether the variable is set or not.
So it seemed consistent/sensible to do the same for the others.
But I am happy to change if you have good suggestions for invalid defaults.

File: .azure-pipelines/templates/build.yml
Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:48]: I think it was also so that this would not be executed on windows

File: .azure-pipelines/templates/build.yml
Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:60]: This only happens temporarily on azure, i think it's fine to have it like that. I think the benefits of not needing a new copy of the env file for each python version outweigh the small inconsistency you point out.

File: .azure-pipelines/templates/build.yml
Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:62]: Sure. But it's already in the job name on the build

File: .azure-pipelines/templates/ccache.yml
Comment (nvaytet) [reply .azure-pipelines/templates/ccache.yml:3]: good idea. I suggest to leave it as is for this PR since it seems to be working, and every time you seem to touch `ccache` it breaks again. We can start using `conda` in a follow-up PR?

File: .azure-pipelines/templates/build.yml
Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:72]: Yes. It depends how long we want the names of the tasks to be.

I also think the name `Python tests and conda build` for the next step is not great either.
Technically it should be `conda build and python tests`.
But i thought it's useful to have `Python tests` at the start of the name because it's actually what you are looking for when you are checking your build, and if you have a narrow browser window you don't see the work Python...

For the same reason, I thought about naming this step just `C++ tests`... I don't know, suggestions welcome!

File: .azure-pipelines/templates/build.yml
Comment (SimonHeybrock) [inline .azure-pipelines/templates/build.yml:14]: In some places you have a parameter named `deploy`, it is a bit confusing which does what. Can there be a better name?

File: .azure-pipelines/templates/deploy.yml
Comment (SimonHeybrock) [inline .azure-pipelines/templates/deploy.yml:14]: `conda_label` to be more explicit?

File: .azure-pipelines/templates/deploy.yml
Comment (SimonHeybrock) [inline .azure-pipelines/templates/deploy.yml:16]: I suppose this should be removed/changed before we merge?

File: tools/make_and_install.sh
Comment (SimonHeybrock) [inline tools/make_and_install.sh:18]: commented code

File: .azure-pipelines/templates/build.yml
Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:14]: `artifact` or `artifact_name` ?

File: .azure-pipelines/templates/deploy.yml
Comment (nvaytet) [reply .azure-pipelines/templates/deploy.yml:16]: i thought it was a good idea to have it as `test`, so that if you mess things up in your setup you don't accidentally push the packages to the main conda label?

Comment (SimonHeybrock) [reply .azure-pipelines/templates/deploy.yml:16]: Not suggestinv that the default should be the main label. Are defaults required? Can we force settings this by not providing a default.

File: .azure-pipelines/templates/build.yml
Comment (SimonHeybrock) [reply .azure-pipelines/templates/build.yml:14]: Artifact of what?

Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:14]: it's the name of the artifact that azure publishes

Comment (SimonHeybrock) [reply .azure-pipelines/templates/build.yml:14]: But where is that name used? For the download links?

Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:14]: What download links?

File: .azure-pipelines/templates/deploy.yml
Comment (nvaytet) [reply .azure-pipelines/templates/deploy.yml:16]: there has to be a default, but it can be an empty string, but i think if you provide an empty label to anaconda upload, it might then default to `main` when uploading

File: .azure-pipelines/templates/build.yml
Comment (SimonHeybrock) [reply .azure-pipelines/templates/build.yml:14]: See, e.g., https://dev.azure.com/scipp/scipp/_build/results?buildId=4447&view=results under "Build and Test". Is this not what you are referring to?

File: .azure-pipelines/templates/deploy.yml
Comment (SimonHeybrock) [reply .azure-pipelines/templates/deploy.yml:16]: Ok, fine with leaving it then 👍

File: .azure-pipelines/templates/build.yml
Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:14]: No. This is what I am refering to: https://dev.azure.com/scipp/scipp/_build/results?buildId=4371&view=artifacts&pathAsName=false&type=publishedArtifacts

But the name is actually created by `conda-build`. When you conda build on linux, it makes a directory called `linux-64`. I get this directory in the `PathtoPublish` and then append the python version to create the `ArtifactName` here: https://github.com/scipp/scipp/blob/81495409c67f4d56146fb1b07a08c4eabed0ea17/.azure-pipelines/templates/build.yml#L86

I think the `publish` name originally came from the `PathtoPublish`.

Comment (SimonHeybrock) [reply .azure-pipelines/templates/build.yml:14]: Ok, basically what I thought then (my link links to your link, I think). `azure_artifact_name`?

Comment (nvaytet) [reply .azure-pipelines/templates/build.yml:14]: But it is not really the full artifact name, it's just the root, or the name of the directory created by conda build.
The artifact name is made up of the `linux-64` (the one we are talking about now) and the python version to make up `linux-64-py3.7`.
Maybe I should just call it `conda_package_root`?

Comment (SimonHeybrock) [reply .azure-pipelines/templates/build.yml:14]: I don't really care, as long as it is clearer what is does when using the template, without having to look at the template in detail.

File: tools/make_and_install.sh
Comment (nvaytet) [reply tools/make_and_install.sh]: What happens with `cmake` when you set the `CMAKE_INSTALL_PREFIX` with an empty string?
`-DCMAKE_INSTALL_PREFIX=""`

File: .azure-pipelines/templates/build.yml
Comment (DanNixon) [reply .azure-pipelines/templates/build.yml:51]: This can just be "not on windows", also not sure why the `OSX_VERSION` variable is being used in the test.

**Review by SimonHeybrock — APPROVED** [review]
Nice work 👍

**Pull request merged**
